File: backend-dotnet/Models/LandTechnicalInspection.cs
// namespace AngularProjectApi.Models;

// public partial class LandTechnicalInspection
// {
//   public int Id { get; set; }

//   public int LandCode { get; set; }

//   public int GovernorateCode { get; set; }

//   public string LandAddress { get; set; } = null!;

//   public string LandOwnerName { get; set; } = null!;

//   public string LandOwnerAddress { get; set; } = null!;

//   public int TotalArea { get; set; }

//   public string LandNature { get; set; } = null!;

//   public int LandOwnershipCode { get; set; }

//   public string? NorthernBoundary { get; set; }

//   public int? NorthernBoundaryLength { get; set; }

//   public string? SouthernBoundary { get; set; }

//   public int? SouthernBoundaryLength { get; set; }

//   public string? EasternBoundary { get; set; }

//   public int? EasternBoundaryLength { get; set; }

//   public string? WesternBoundary { get; set; }

//   public int? WesternBoundaryLength { get; set; }

//   public string? NortheastBoundary { get; set; }

//   public int? NortheastBoundaryLength { get; set; }

//   public string? NorthwestBoundary { get; set; }

//   public int? NorthwestBoundaryLength { get; set; }

//   public string? SoutheastBoundary { get; set; }

//   public int? SoutheastBoundaryLength { get; set; }

//   public string? SouthwestBoundary { get; set; }

//   public int? SouthwestBoundaryLength { get; set; }

//   public Guid? TechnicalResponsiblePersonId { get; set; }

//   public int? LegalResponsiblePersonId { get; set; }

//   public DateTime? NeedsCommitteeDate { get; set; }

//   public DateTime? TechnicalInspectionDate { get; set; }
//   public DateTime? CreatedAt { get; set; }
// }



using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AngularProjectApi.Models;

[Table("LandTechnicalInspection")]
public partial class LandTechnicalInspection
{
    [Key]
    public int Id { get; set; }

    [Required(ErrorMessage = "كود الأرض مطلوب")]
    public int LandCode { get; set; }

    public virtual Land? Land { get; set; }

    public ICollection<LandAndLegalConnection> LandAndLegalConnections { get; set; } = new List<LandAndLegalConnection>();

    [Required(ErrorMessage = "كود المحافظة مطلوب")]
    public int GovernorateCode { get; set; }

    [Required(ErrorMessage = "عنوان الأرض مطلوب")]
    [StringLength(500, ErrorMessage = "العنوان طويل جداً")]
    public string LandAddress { get; set; } = null!;

    [Required(ErrorMessage = "اسم مالك الأرض مطلوب")]
    [StringLength(200, MinimumLength = 3, ErrorMessage = "الاسم يجب أن يكون 3 أحرف على الأقل")]
    public string LandOwnerName { get; set; } = null!;

    [Required(ErrorMessage = "عنوان مالك الأرض مطلوب")]
    public string LandOwnerAddress { get; set; } = null!;

    [Required(ErrorMessage = "المساحة الإجمالية مطلوبة")]
    [Range(1, int.MaxValue, ErrorMessage = "المساحة يجب أن تكون أكبر من صفر")]
    public int TotalArea { get; set; }

    [Required(ErrorMessage = "طبيعة الأرض مطلوبة")]
    public string LandNature { get; set; } = null!;

    [Required(ErrorMessage = "كود ملكية الأرض مطلوب")]
    public int LandOwnershipCode { get; set; }

    [Required(ErrorMessage = "الحد الشمالي مطلوب")]
    public string? NorthernBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الشمالي مطلوب")]
    [Range(0, int.MaxValue, ErrorMessage = "الطول لا يمكن أن يكون سالباً")]
    public int? NorthernBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الجنوبي مطلوب")]
    public string? SouthernBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الجنوبي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? SouthernBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الشرقي مطلوب")]
    public string? EasternBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الشرقي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? EasternBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الغربي مطلوب")]
    public string? WesternBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الغربي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? WesternBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الشمالي الشرقي مطلوب")]
    public string? NortheastBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الشمالي الشرقي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? NortheastBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الشمالي الغربي مطلوب")]
    public string? NorthwestBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الشمالي الغربي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? NorthwestBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الجنوبي الشرقي مطلوب")]
    public string? SoutheastBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الجنوبي الشرقي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? SoutheastBoundaryLength { get; set; }

    [Required(ErrorMessage = "الحد الجنوبي الغربي مطلوب")]
    public string? SouthwestBoundary { get; set; }

    [Required(ErrorMessage = "طول الحد الجنوبي الغربي مطلوب")]
    [Range(0, int.MaxValue)]
    public int? SouthwestBoundaryLength { get; set; }

    [Required(ErrorMessage = "المسؤول الفني مطلوب")]
    public Guid? TechnicalResponsiblePersonId { get; set; }

    [Required(ErrorMessage = "المسؤول القانوني مطلوب")]
    public int? LegalResponsiblePersonId { get; set; }

    [Required(ErrorMessage = "تاريخ احتياج اللجنة مطلوب")]
    [DataType(DataType.Date)]
    public DateTime? NeedsCommitteeDate { get; set; }

    [Required(ErrorMessage = "تاريخ المعاينة الفنية مطلوب")]
    [DataType(DataType.Date)]
    public DateTime? TechnicalInspectionDate { get; set; }

    [Required]
    public DateTime? CreatedAt { get; set; } = DateTime.Now;
}

File: frontend-angular/src/app/components/database-operations/land-technical-inspection/land-technical-inspection.ts
import { Component, signal, inject } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule, FormBuilder, FormGroup, Validators } from '@angular/forms';
import { HeaderComponent } from '../../shared/header/header';
import { Router } from '@angular/router';
import { LookupApiService } from '../../../services/lookup-api.service';
import { GovernorateData, LandOwnerData } from '../../../models/Lookup.Model';
import { LandTechnicalInspection } from '../../../models/landtechnicalinspection.model';
import { ErrorHandlerService } from '../../../services/error-handler.service';
import { LandTechnicalInspectionApiService } from '../../../services/land-technical-inspection-api.service';
import { LandApiService } from '../../../services/land-api.service';

type ViewMode = 'search' | 'view' | 'create' | 'edit';

@Component({
  selector: 'app-land-technical-inspection',
  standalone: true,
  imports: [CommonModule, FormsModule, ReactiveFormsModule, HeaderComponent],
  templateUrl: './land-technical-inspection.html',
  styleUrl: './land-technical-inspection.css'
})

export class LandTechnicalInspectionComponent {

  private router = inject(Router);
  private lookupApiService = inject(LookupApiService);
  private errorHandler = inject(ErrorHandlerService);
  private landTechnicalInspectionApiService = inject(LandTechnicalInspectionApiService);
  private landApiService = inject(LandApiService);

  selectedLandTechnicalInspection = signal<LandTechnicalInspection | null>(null);
  viewMode = signal<ViewMode>('search');
  showDialog = signal<boolean>(false);
  dialogTitle = signal<string>('');
  dialogMessage = signal<string>('');
  dialogVariant = signal<'success' | 'error'>('success');
  private dialogRedirectToHomeAfterClose = false;
  technicalResponsiblePersonNameAndId: string;
  loggedInUser = JSON.parse(localStorage.getItem("currentUser") || '""');

  //Lookup
  governorates: GovernorateData[] = []; // Declare an empty array of MyObject
  landOwners: LandOwnerData[] = []; // Declare an empty array of MyObject
  landCodes: number[] = [];
  landCodeNotFound = false;

  get boundariesGroup() {
    return this.form.get('boundaries') as FormGroup;
  }
  get lengthsGroup() {
    return this.form.get('lengths') as FormGroup;
  }
  private fb = inject(FormBuilder);

  protected goBack(): void {
    this.router.navigate(['/property-registration-list-insertions']);
  }

  protected goHome(): void {
    this.router.navigate(['/dashboard']);
  }

  protected logout(): void {
    this.router.navigate(['/login']);
  }

  protected closeDialog(): void {
    this.showDialog.set(false);

    if (this.dialogRedirectToHomeAfterClose) {
      this.dialogRedirectToHomeAfterClose = false;
      this.goHome();
    }
  }

  private openDialog(variant: 'success' | 'error', title: string, message: string, redirectToHomeAfterClose = false): void {
    this.dialogVariant.set(variant);
    this.dialogTitle.set(title);
    this.dialogMessage.set(message);
    this.dialogRedirectToHomeAfterClose = redirectToHomeAfterClose;
    this.showDialog.set(true);
  }

  constructor() {
    this.loadGovernorates();
    this.loadLandOwners();
    this.loadLandCodes();

    this.technicalResponsiblePersonNameAndId = this.loggedInUser.id + ' ' + this.loggedInUser.fullName
  }

  form: FormGroup = this.fb.group({
    landCode: ['', [Validators.required, Validators.min(1)]],
    governorateCode: ['', [Validators.required, Validators.min(1)]],
    landAddress: ['', [Validators.required, Validators.maxLength(500)]],
    landOwnerName: ['', [Validators.required, Validators.minLength(3), Validators.maxLength(200)]],
    landOwnerAddress: ['', Validators.required],
    totalArea: ['', [Validators.required, Validators.min(1)]],
    landNature: ['مباني', Validators.required],
    landOwnershipCode: ['', [Validators.required, Validators.min(1)]],

    // الحدود
    northernBoundary: ['', Validators.required],
    southernBoundary: ['', Validators.required],
    easternBoundary: ['', Validators.required],
    westernBoundary: ['', Validators.required],

    // الأطوال
    northernBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    southernBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    easternBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    westernBoundaryLength: ['', [Validators.required, Validators.min(1)]],

    // الزوايا
    northeastBoundary: ['', Validators.required],
    northwestBoundary: ['', Validators.required],
    southeastBoundary: ['', Validators.required],
    southwestBoundary: ['', Validators.required],

    // أطوال الزوايا
    northeastBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    northwestBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    southeastBoundaryLength: ['', [Validators.required, Validators.min(1)]],
    southwestBoundaryLength: ['', [Validators.required, Validators.min(1)]],

    technicalResponsiblePersonId: [this.loggedInUser.id, Validators.required],
    legalResponsiblePersonId: ['', [Validators.required, Validators.min(1)]],
    needsCommitteeDate: ['', Validators.required],
    technicalInspectionDate: ['', Validators.required]
  });

  private loadLandCodes(): void {
    this.landApiService.getAvailableLandCodes().subscribe({
      next: (codes) => {
        this.landCodes = codes || [];
      },
      error: (error) => {
        const errorMessage = this.errorHandler.getUserFriendlyMessage(error, 'تحميل أكواد الأراضي المتاحة');
        this.openDialog('error', 'خطأ', errorMessage);
      }
    });
  }

  protected onLandCodeChanged(): void {
    const raw = this.form.get('landCode')?.value;
    const landCode = Number(raw);
    if (!landCode || Number.isNaN(landCode)) {
      this.landCodeNotFound = false;
      return;
    }

    if (this.landCodes.length > 0 && !this.landCodes.includes(landCode)) {
      this.landCodeNotFound = true;
      return;
    }

    this.landCodeNotFound = false;
  }

  private applyServerValidationErrors(error: any): void {
    const errors = error?.error?.errors;
    if (!errors || typeof errors !== 'object') {
      return;
    }

    Object.keys(errors).forEach((key) => {
      const messages = Array.isArray(errors[key]) ? errors[key] : [errors[key]];
      const message = messages.find((m: any) => typeof m === 'string' && m.trim().length > 0) ?? null;
      if (!message) {
        return;
      }

      const normalizedKey = (key ?? '').toString();
      const candidateNames = [
        normalizedKey,
        normalizedKey.length > 0 ? normalizedKey[0].toLowerCase() + normalizedKey.slice(1) : '',
        normalizedKey.toLowerCase()
      ].filter(Boolean);

      const formControls = Object.keys(this.form.controls);
      const matchedControlName = formControls.find(c => candidateNames.some(k => k === c || k.toLowerCase() === c.toLowerCase()));
      if (!matchedControlName) {
        return;
      }

      const control = this.form.get(matchedControlName);
      if (!control) {
        return;
      }

      control.setErrors({ ...(control.errors ?? {}), server: message });
      control.markAsTouched();
    });
  }

  private loadLandOwners() {
    this.lookupApiService.getLandOwners().subscribe({
      next: (landOwnersList) => {
        // Guard against null and normalize the response to an array
        if (!landOwnersList) {
          this.landOwners = [];
          return;
        }
        this.landOwners = Array.isArray(landOwnersList) ? landOwnersList : [landOwnersList];
      },
      error: (error) => {
        const errorMessage = this.errorHandler.getUserFriendlyMessage(
          error,
          'تحميل أصحاب الأراضي'
        );
        this.openDialog('error', 'خطأ', errorMessage);
      }
    });
  }

  private loadGovernorates() {
    this.lookupApiService.getGovernorates().subscribe({
      next: (governoratesList) => {
        // Guard against null and normalize the response to an array
        if (!governoratesList) {
          this.governorates = [];
          return;
        }
        this.governorates = Array.isArray(governoratesList) ? governoratesList : [governoratesList];
      },
      error: (error) => {
        const errorMessage = this.errorHandler.getUserFriendlyMessage(
          error,
          'تحميل المحافظات'
        );
        this.openDialog('error', 'خطأ', errorMessage);
      }
    });
  }

  protected submit(): void {
    this.form.setErrors(null);

    Object.keys(this.form.controls).forEach((k) => {
      const c = this.form.get(k);
      if (!c || !c.errors || !c.errors['server']) {
        return;
      }
      const { server, ...rest } = c.errors;
      c.setErrors(Object.keys(rest).length > 0 ? rest : null);
    });

    this.onLandCodeChanged();
    if (this.landCodeNotFound) {
      this.form.get('landCode')?.setErrors({ ...(this.form.get('landCode')?.errors ?? {}), server: 'لا توجد قطعة أرض بهذا الكود في جدول الأراضي.' });
      this.form.get('landCode')?.markAsTouched();
      return;
    }

    if (this.form.invalid) {
      this.form.markAllAsTouched();
      return;
    }

    this.landTechnicalInspectionApiService.saveLandTechnicalInspection(this.form.value).subscribe({
      next: (landTechnicalInspection) => {
        this.selectedLandTechnicalInspection.set(landTechnicalInspection);
        this.viewMode.set('view');

        // Send form data to backend -- currently logging and clearing form
        console.log('Land inspection data submitted', this.form.value);
        this.form.reset();
        this.openDialog('success', 'تم الحفظ', 'تم حفظ بيانات المعاينة الفنية للأرض بنجاح!', true);
      },
      error: (error) => {
        console.error('Error creating building:', error);

        this.applyServerValidationErrors(error);

        const errorMessage = this.errorHandler.getUserFriendlyMessage(
          error,
          'إضافة معاينة فنية للأرض'
        );
        this.openDialog('error', 'فشل الحفظ', `فشل في إضافة معاينة فنية للأرض:\n${errorMessage}`);
      }
    });
  }
}